src/bios/flib/src/fs/ext4.rs

```diff
@@ -0,0 +1,671 @@
+
+        let inode : &Inode;
+        let inode_addr = (0x1500 + inode_address % 512) as *mut Inode;
+
+        debug!(0x1500 + inode_address % 512);
+
+        inode = unsafe {
+            transmute(inode_addr)
+        };
+
+        return inode;
+    }
+}
+
+#[repr(C, packed)]
+struct BlockGroupDescriptor32 {
+    bg_block_bitmap : u32,
+    bg_inode_bitmap : u32,
+    bg_inode_table : u32,
+    bg_free_blocks_count : u16,
+    bg_free_inodes_count : u16,
+    bg_used_dirs_count : u16,
+    bg_flags : u16,
+    bg_exclude_bitmap_lo : u32,
+    bg_block_bitmap_csum_lo : u16,
+    bg_inode_bitmap_csum_lo : u16,
+    bg_itable_unused_lo : u16,
+    bg_checksum : u16,
+    bg_reserved : [u8; 32]
+}
+
+#[repr(C, packed)]
+pub struct Inode {
+    pub i_mode : u16,
+    i_uid : u16,
+    i_size : u32,
+    i_atime : u32,
+    i_ctime : u32,
+    i_mtime : u32,
+    i_dtime : u32,
+    i_gid : u16,
+    i_links_count : u16,
+    pub i_blocks : u32,
+    i_flags : u32,
+    i_osd1 : u32,
+    pub i_block : [u32; 15],
+    i_generation : u32,
+    i_file_acl : u32,
+    i_dir_acl : u32,
+    i_faddr : u32,
+    i_osd2 : [u8; 12],
+    i_extra_isize : u16,
+    i_checksum_hi : u16,
+    i_ctime_extra : u32,
+    i_mtime_extra : u32,
+    i_atime_extra : u32,
+    i_crtime : u32,
+    i_crtime_extra : u32,
+    i_version_hi : u32,
+    i_projid : u32,
+}
+
+#[repr(C, packed)]
+struct Ext4ExtentHeader {
+    eh_magic : u16,
+    eh_entries : u16,
+    eh_max : u16,
+    eh_depth : u16,
+    eh_generation : u32
+}
+
+#[repr(C, packed)]
+struct Ext4Extent {
+    ee_block : u32,
+    ee_len : u16,
+    ee_start_hi : u16,
+    ee_start_lo : u32
+}
+
+#[repr(C, packed)]
+struct Ext4ExtentIdx {
+    ei_block : u32,
+    ei_leaf_lo : u32,
+    ei_leaf_hi : u16,
+    ei_unused : u16
+}
+
+
+impl Inode {
+    // Returns true is this inode uses an extent tree
+    pub fn uses_extent_tree(&self) -> bool {
+        return self.i_flags == 0x80000
+    }
+
+    // Get block number of the nth data block.
+    pub fn get_nth_data_block(&self, block_size : u32, n : u32, partition : &Ext4Partition) -> u64 {
+        let q = self.i_size as i32 / block_size as i32;
+        let r = self.i_size as i32 % block_size as i32;
+        if ! ( (n <= q as u32) | ( ( n == (q as u32 + 1)) & (r > 0) ))  {
+            return 0;
+        }
+        if self.uses_extent_tree() {
+            print_str("Calling extent");
+            return self.get_nth_data_block_extent(block_size, n, partition);
+        } else {
+            print_str("No extent tree");
+            return 0
+        }
+    }
+
+    // Get block number of the nth data block, considering this inode uses an extent tree structure
+    pub fn get_nth_data_block_extent(&self, block_size : u32, n : u32, partition : &Ext4Partition) -> u64 {
+        let mem_offset = self as *const Inode as u32 + 0x28;
+        debug!(mem_offset);
+        return self.explore_next_layer(mem_offset , n , block_size, partition);
```

Comment:
```suggestion
        self.explore_next_layer(mem_offset , n, block_size, partition)
```
si tu préfères laisser les return, tu peux